daemon/cmd/ipcache.go

```diff
@@ -87,3 +104,128 @@ func containsSubnet(outer, inner net.IPNet) bool {
 
 	return outerBits == innerBits && outerOnes <= innerOnes && outer.Contains(inner.IP)
 }
+
+// restoreIPCache dumps the existing (old) bpf ipcache, adding relevant information
+// back in to the ipcache / identity allocator.
+//
+// For all local (cidr / remote-node) identities, it will add a stub metadata entry
+// that ensures the prefix exists and it gets the same numeric identity before. It
+// reserves those local numeric IDs from allocation to ensure they are available.
+//
+// For ingress IPs, it will add those to the ipcache and configure the local node
+// accordingly.
+//
+// This *must* be called before initMaps(), which will hide the "old" ipcache.
+func (d *Daemon) restoreIPCache() error {
+	ingressIPs := make([]netip.Prefix, 0, 2)
+	// need to preserve this so we can remove it later.
+	d.restoredCIDRs = map[netip.Prefix]identity.NumericIdentity{}
+
+	// Dump the bpf ipcache, recording any prefixes with local or ingress
+	// numeric identities.
+	err := ipcachemap.IPCacheMap().DumpWithCallback(func(key bpf.MapKey, value bpf.MapValue) {
+		k := key.(*ipcachemap.Key)
+		v := value.(*ipcachemap.RemoteEndpointInfo)
+		nid := identity.NumericIdentity(v.SecurityIdentity)
+
+		if isLocalIdentity(nid) {
+			d.restoredCIDRs[k.Prefix()] = nid
+		} else if nid == identity.ReservedIdentityIngress && v.TunnelEndpoint.IsZero() {
+			ingressIPs = append(ingressIPs, k.Prefix())
+		}
+	})
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("error dumping ipcache: %w", err)
+	}
+	// DumpWithCallback() leaves the ipcache map open, must close before opened for
+	// parallel mode in Daemon.initMaps()
+	ipcachemap.IPCacheMap().Close()
```

Comment:
potential bug: The old version of the code did log on error, while the new code does an early return if `DumpWithCallback` failed.  And given that the caller continues on error, I think this violates the comment here which says that close is important.

I think we want to close this irrespective of the `err`, i.e. move to before the `if` statement (or perhaps `defer` it)

Reply:
Oh, very good catch.